### PR TITLE
hypre: fix sycl build 

### PR DESCRIPTION
**Pull request by osborn9**
This fixes the SYCL variant build when oneapi dependencies are installed via spack.

### CODE REVIEW

File: var/spack/repos/builtin/packages/hypre/package.py [hunk 1]
```diff
@@ -279,13 +281,18 @@ def configure_args(self):
...
             if not (os.path.basename(self.compiler.cxx) in sycl_compatible_compilers):
                 raise InstallError(
                     "Hypre's SYCL GPU Backend requires DPC++ (dpcpp)"
                     + " or the oneAPI CXX (icpx) compiler."
                 )
+            sycl_inc = ""
+            sycl_pkgs = ["intel-oneapi-dpl", "intel-oneapi-mkl"]
+            for pkg in sycl_pkgs:
+                if "^" + pkg in spec:
+                    sycl_inc += spec[pkg].headers.include_flags + " "
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/hypre/package.py:298]: It seems to me both dependencies are enforced when we activate `+sycl`, so this:
```suggestion
            for pkg in ("intel-oneapi-dpl", "intel-oneapi-mkl")
                sycl_inc += spec[pkg].headers.include_flags + " "
```
should be enough, right?

**Review by alalazo — CHANGES_REQUESTED** [review]

**Comment by eugeneswalker** [comment]
Trying to use this PR to build `hypre +sycl` still fails for me, although it seems to go further than it did before the changes in this PR.

Any idea what is going on here @osborn9 ? I take it you did not run into this.

```
$> docker run -it ecpe4s/ubuntu22.04-runner-amd64-oneapi-2023.2.1-mpi-base:2023.10.19

root@cd0c42d6e0b7:/# icx --version
Intel(R) oneAPI DPC++/C++ Compiler 2023.2.0 (2023.2.0.20230721)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/intel/oneapi/compiler/2023.2.1/linux/bin-llvm
Configuration file: /opt/intel/oneapi/compiler/2023.2.1/linux/bin-llvm/../bin/icx.cfg

$> cat <<EOF >spack.yaml
spack:
  view: false

  concretizer:
    reuse: false
    unify: false

  compilers:
  - compiler:
      spec: oneapi@2023.2.1
      paths:
        cc: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/icx
        cxx: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/icpx
        f77: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/ifx
        fc: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/ifx
      flags: {}
      operating_system: ubuntu22.04
      target: x86_64
      modules: []
      environment:
        append_path:
          PATH: /opt/intel/oneapi/compiler/2023.2.1/linux/bin-llvm
      extra_rpaths: []

  packages:
    all:
      require: '%oneapi'
      providers:
        mpi: [mpich]
        tbb: [intel-tbb]
      target: [x86_64]
      variants: +mpi
    ncurses:
      variants: +termlib
    xz:
      variants: +pic
    mpi:
      require: mpich
    mpich:
      require: '@4.1.2 ~wrapperrpath ~hwloc'
      buildable: false
      externals:
      - spec: mpich@4.1.2 %oneapi@2023.2.1 ~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi
        prefix: /usr/local/mpich/install/mpich

  specs:
  - hypre +sycl
EOF

$> spack -e . concretize -f | tee concretize.log
 -   5yhhngk  hypre@2.29.0%oneapi@2023.2.1~caliper~complex~cuda~debug+fortran~gptune~int64~internal-superlu~magma~mixedint+mpi~openmp~rocm+shared~superlu-dist+sycl~umpire~unified-memory build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]  dxwi74k      ^gmake@4.4.1%oneapi@2023.2.1~guile build_system=generic arch=linux-ubuntu22.04-x86_64
[+]  22yjvcy      ^intel-oneapi-dpl@2022.2.0%oneapi@2023.2.1+envmods build_system=generic arch=linux-ubuntu22.04-x86_64
[+]  pkbadrx      ^intel-oneapi-mkl@2023.2.0%oneapi@2023.2.1~cluster+envmods~ilp64+shared build_system=generic mpi_family=none threads=none arch=linux-ubuntu22.04-x86_64
[+]  r5zht4j          ^intel-tbb@2021.9.0%oneapi@2023.2.1~ipo+shared+tm build_system=cmake build_type=Release cxxstd=default generator=make patches=91755c6 arch=linux-ubuntu22.04-x86_64
[+]  6uyskok              ^cmake@3.27.7%oneapi@2023.2.1~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu22.04-x86_64
[+]  pk2extv                  ^curl@8.4.0%oneapi@2023.2.1~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu22.04-x86_64
[+]  6s5yqqz                      ^nghttp2@1.57.0%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]  kuy5us6                      ^openssl@3.1.3%oneapi@2023.2.1~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-x86_64
[+]  qqkmghd                          ^ca-certificates-mozilla@2023-05-30%oneapi@2023.2.1 build_system=generic arch=linux-ubuntu22.04-x86_64
[+]  tqg54fh                          ^perl@5.38.0%oneapi@2023.2.1+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-x86_64
[+]  edb5alc                              ^berkeley-db@18.1.40%oneapi@2023.2.1+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-x86_64
[+]  rlbxfsa                              ^bzip2@1.0.8%oneapi@2023.2.1~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-x86_64
[+]  6vysvyp                              ^gdbm@1.23%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]  c2jddi3                                  ^readline@8.2%oneapi@2023.2.1 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-x86_64
[+]  2z5gdjw                  ^ncurses@6.4%oneapi@2023.2.1~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]  s24tzln                  ^zlib-ng@2.1.4%oneapi@2023.2.1+compat+opt build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]  hw2d35t              ^hwloc@2.9.1%oneapi@2023.2.1~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64
[+]  d4eivx4                  ^libpciaccess@0.17%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]  62x5f5w                      ^libtool@2.4.7%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]  rq7pvm7                          ^m4@1.4.19%oneapi@2023.2.1+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu22.04-x86_64
[+]  kcdxwqy                              ^diffutils@3.9%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]  73j6tgz                              ^libsigsegv@2.14%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]  imb276g                      ^util-macros@1.19.3%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]  lvrfoib                  ^libxml2@2.10.3%oneapi@2023.2.1+pic~python+shared build_system=autotools arch=linux-ubuntu22.04-x86_64
[+]  zvzqfui                      ^libiconv@1.17%oneapi@2023.2.1 build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64
[+]  imfndd7                      ^xz@5.4.1%oneapi@2023.2.1+pic build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64
[+]  fw5fyh2                  ^pkgconf@1.9.5%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
[e]  x3bdfob      ^mpich@4.1.2%oneapi@2023.2.1~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu22.04-x86_64

$> spack -e . install
...
==> Installing hypre-2.29.0-5yhhngkhedypyodxdqvved7wmfovtz7l [13/13]
==> No binary for hypre-2.29.0-5yhhngkhedypyodxdqvved7wmfovtz7l found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/98/98b72115407a0e24dbaac70eccae0da3465f8f999318b2c9241631133f42d511.tar.gz
==> No patches needed for hypre
==> hypre: Executing phase: 'autoreconf'
==> hypre: Executing phase: 'configure'
==> hypre: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

3 errors found in build log:
     369                                         ^
     370    4 warnings generated.
     371    Building libHYPRE_utilities.so ...
     372    /usr/local/mpich/install/mpich/bin/mpicc -shared  -o libHYPRE_utilities-2.29.0.so F90_HYPRE_error.o F90_HYPRE_general.o HYPRE_handle.o HYPRE_version.o amg_linklist.o binsearc
            h.o exchange_data.o fortran_matrix.o ap.o log.o complex.o error.o int_array.o hopscotch_hash.o magma.o memory_tracker.o merge_sort.o mmio.o mpi_comm_f2c.o prefix_sum.o printf
            .o qsort.o utilities.o mpistubs.o qsplit.o random.o state.o threading.o timer.o timing.o device_utils.obj general.obj handle.obj int_array_device.obj memory.obj omp_device.ob
            j nvtx.obj -Wl,-soname,libHYPRE_utilities-2.29.0.so
     373    /usr/bin/ld: device_utils.obj: relocation R_X86_64_32S against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
     374    /usr/bin/ld: failed to set dynamic section sizes: bad value
  >> 375    icx: error: linker command failed with exit code 1 (use -v to see invocation)
  >> 376    make[1]: *** [Makefile:121: libHYPRE_utilities.so] Error 1
     377    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-hypre-2.29.0-5yhhngkhedypyodxdqvved7wmfovtz7l/spack-src/src/utilities'
  >> 378    make: *** [Makefile:86: all] Error 1
```

**Comment by osborn9** [comment]
> Trying to use this PR to build `hypre +sycl` still fails for me, although it seems to go further than it did before the changes in this PR.
> 
> Any idea what is going on here @osborn9 ? I take it you did not run into this.
> 
> ```
> $> docker run -it ecpe4s/ubuntu22.04-runner-amd64-oneapi-2023.2.1-mpi-base:2023.10.19
> 
> root@cd0c42d6e0b7:/# icx --version
> Intel(R) oneAPI DPC++/C++ Compiler 2023.2.0 (2023.2.0.20230721)
> Target: x86_64-unknown-linux-gnu
> Thread model: posix
> InstalledDir: /opt/intel/oneapi/compiler/2023.2.1/linux/bin-llvm
> Configuration file: /opt/intel/oneapi/compiler/2023.2.1/linux/bin-llvm/../bin/icx.cfg
> 
> $> cat <<EOF >spack.yaml
> spack:
>   view: false
> 
>   concretizer:
>     reuse: false
>     unify: false
> 
>   compilers:
>   - compiler:
>       spec: oneapi@2023.2.1
>       paths:
>         cc: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/icx
>         cxx: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/icpx
>         f77: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/ifx
>         fc: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/ifx
>       flags: {}
>       operating_system: ubuntu22.04
>       target: x86_64
>       modules: []
>       environment:
>         append_path:
>           PATH: /opt/intel/oneapi/compiler/2023.2.1/linux/bin-llvm
>       extra_rpaths: []
> 
>   packages:
>     all:
>       require: '%oneapi'
>       providers:
>         mpi: [mpich]
>         tbb: [intel-tbb]
>       target: [x86_64]
>       variants: +mpi
>     ncurses:
>       variants: +termlib
>     xz:
>       variants: +pic
>     mpi:
>       require: mpich
>     mpich:
>       require: '@4.1.2 ~wrapperrpath ~hwloc'
>       buildable: false
>       externals:
>       - spec: mpich@4.1.2 %oneapi@2023.2.1 ~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi
>         prefix: /usr/local/mpich/install/mpich
> 
>   specs:
>   - hypre +sycl
> EOF
> 
> $> spack -e . concretize -f | tee concretize.log
>  -   5yhhngk  hypre@2.29.0%oneapi@2023.2.1~caliper~complex~cuda~debug+fortran~gptune~int64~internal-superlu~magma~mixedint+mpi~openmp~rocm+shared~superlu-dist+sycl~umpire~unified-memory build_system=autotools arch=linux-ubuntu22.04-x86_64
> [+]  dxwi74k      ^gmake@4.4.1%oneapi@2023.2.1~guile build_system=generic arch=linux-ubuntu22.04-x86_64
> [+]  22yjvcy      ^intel-oneapi-dpl@2022.2.0%oneapi@2023.2.1+envmods build_system=generic arch=linux-ubuntu22.04-x86_64
> [+]  pkbadrx      ^intel-oneapi-mkl@2023.2.0%oneapi@2023.2.1~cluster+envmods~ilp64+shared build_system=generic mpi_family=none threads=none arch=linux-ubuntu22.04-x86_64
> [+]  r5zht4j          ^intel-tbb@2021.9.0%oneapi@2023.2.1~ipo+shared+tm build_system=cmake build_type=Release cxxstd=default generator=make patches=91755c6 arch=linux-ubuntu22.04-x86_64
> [+]  6uyskok              ^cmake@3.27.7%oneapi@2023.2.1~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu22.04-x86_64
> [+]  pk2extv                  ^curl@8.4.0%oneapi@2023.2.1~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu22.04-x86_64
> [+]  6s5yqqz                      ^nghttp2@1.57.0%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
> [+]  kuy5us6                      ^openssl@3.1.3%oneapi@2023.2.1~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-x86_64
> [+]  qqkmghd                          ^ca-certificates-mozilla@2023-05-30%oneapi@2023.2.1 build_system=generic arch=linux-ubuntu22.04-x86_64
> [+]  tqg54fh                          ^perl@5.38.0%oneapi@2023.2.1+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-x86_64
> [+]  edb5alc                              ^berkeley-db@18.1.40%oneapi@2023.2.1+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-x86_64
> [+]  rlbxfsa                              ^bzip2@1.0.8%oneapi@2023.2.1~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-x86_64
> [+]  6vysvyp                              ^gdbm@1.23%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
> [+]  c2jddi3                                  ^readline@8.2%oneapi@2023.2.1 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-x86_64
> [+]  2z5gdjw                  ^ncurses@6.4%oneapi@2023.2.1~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu22.04-x86_64
> [+]  s24tzln                  ^zlib-ng@2.1.4%oneapi@2023.2.1+compat+opt build_system=autotools arch=linux-ubuntu22.04-x86_64
> [+]  hw2d35t              ^hwloc@2.9.1%oneapi@2023.2.1~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64
> [+]  d4eivx4                  ^libpciaccess@0.17%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
> [+]  62x5f5w                      ^libtool@2.4.7%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
> [+]  rq7pvm7                          ^m4@1.4.19%oneapi@2023.2.1+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu22.04-x86_64
> [+]  kcdxwqy                              ^diffutils@3.9%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
> [+]  73j6tgz                              ^libsigsegv@2.14%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
> [+]  imb276g                      ^util-macros@1.19.3%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
> [+]  lvrfoib                  ^libxml2@2.10.3%oneapi@2023.2.1+pic~python+shared build_system=autotools arch=linux-ubuntu22.04-x86_64
> [+]  zvzqfui                      ^libiconv@1.17%oneapi@2023.2.1 build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64
> [+]  imfndd7                      ^xz@5.4.1%oneapi@2023.2.1+pic build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64
> [+]  fw5fyh2                  ^pkgconf@1.9.5%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu22.04-x86_64
> [e]  x3bdfob      ^mpich@4.1.2%oneapi@2023.2.1~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu22.04-x86_64
> 
> $> spack -e . install
> ...
> ==> Installing hypre-2.29.0-5yhhngkhedypyodxdqvved7wmfovtz7l [13/13]
> ==> No binary for hypre-2.29.0-5yhhngkhedypyodxdqvved7wmfovtz7l found: installing from source
> ==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/98/98b72115407a0e24dbaac70eccae0da3465f8f999318b2c9241631133f42d511.tar.gz
> ==> No patches needed for hypre
> ==> hypre: Executing phase: 'autoreconf'
> ==> hypre: Executing phase: 'configure'
> ==> hypre: Executing phase: 'build'
> ==> Error: ProcessError: Command exited with status 2:
>     'make' '-j16'
> 
> 3 errors found in build log:
>      369                                         ^
>      370    4 warnings generated.
>      371    Building libHYPRE_utilities.so ...
>      372    /usr/local/mpich/install/mpich/bin/mpicc -shared  -o libHYPRE_utilities-2.29.0.so F90_HYPRE_error.o F90_HYPRE_general.o HYPRE_handle.o HYPRE_version.o amg_linklist.o binsearc
>             h.o exchange_data.o fortran_matrix.o ap.o log.o complex.o error.o int_array.o hopscotch_hash.o magma.o memory_tracker.o merge_sort.o mmio.o mpi_comm_f2c.o prefix_sum.o printf
>             .o qsort.o utilities.o mpistubs.o qsplit.o random.o state.o threading.o timer.o timing.o device_utils.obj general.obj handle.obj int_array_device.obj memory.obj omp_device.ob
>             j nvtx.obj -Wl,-soname,libHYPRE_utilities-2.29.0.so
>      373    /usr/bin/ld: device_utils.obj: relocation R_X86_64_32S against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
>      374    /usr/bin/ld: failed to set dynamic section sizes: bad value
>   >> 375    icx: error: linker command failed with exit code 1 (use -v to see invocation)
>   >> 376    make[1]: *** [Makefile:121: libHYPRE_utilities.so] Error 1
>      377    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-hypre-2.29.0-5yhhngkhedypyodxdqvved7wmfovtz7l/spack-src/src/utilities'
>   >> 378    make: *** [Makefile:86: all] Error 1
> ```

Yes, this should be resolved with https://github.com/hypre-space/hypre/pull/995.

**Comment by eugeneswalker** [comment]
> Yes, this should be resolved with [hypre-space/hypre#995](https://github.com/hypre-space/hypre/pull/995).

Should we make this commit a patch and tell Spack to apply it to `hypre@2.29.0 +sycl` in this PR?

**Comment by balay** [comment]
@osborn9 I get the following build fauilre for my `hypre+sycl` build (on hypre/master branch)

```
icpx  -fPIC -O3 -fsycl -fsycl-unnamed-lambda -I/software/intel/oneapi/dpl/2022.2.0/linux/include -I/software/intel/oneapi/mkl/2023.2.0/include  -DHAVE_CONFIG_H -I.. -I./.. -I./../struct_mv -I.    -qmkl -I/include -DMKL_ILP64 -I/software/intel/oneapi/mkl/2023.2.0/include      -I/software/intel/oneapi/include  -c device_utils.c -o device_utils.obj
...
In file included from device_utils.c:8:
In file included from ./_hypre_onedpl.hpp:18:
In file included from /software/intel/oneapi/dpl/2022.2.0/linux/include/oneapi/dpl/execution:65:
In file included from /software/intel/oneapi/dpl/2022.2.0/linux/include/oneapi/dpl/pstl/algorithm_impl.h:31:
In file included from /software/intel/oneapi/dpl/2022.2.0/linux/include/oneapi/dpl/pstl/parallel_backend.h:26:
/software/intel/oneapi/dpl/2022.2.0/linux/include/oneapi/dpl/pstl/parallel_backend_tbb.h:28:10: fatal error: 'tbb/blocked_range.h' file not found
#include <tbb/blocked_range.h>
         ^~~~~~~~~~~~~~~~~~~~~
```

[spack-build-out.txt](https://github.com/spack/spack/files/13253757/spack-build-out.txt)

File: var/spack/repos/builtin/packages/hypre/package.py (hunk 1 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/hypre/package.py:298]: Is the PR still a work in progress?

**Pull request closed without merging**